### PR TITLE
Buffer re-use using ArrayPool and a few more checks

### PR DESCRIPTION
**Pull request by harshithapv**
Added ArrayPool for buffer re-use while reading images in ImageLoader.cs. A few commits for safety checks. 
Continuation to my previous commit #4242

### CODE REVIEW

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 1]
```diff
@@ -275,6 +275,8 @@ public Tensor ProcessImage(in VBuffer<byte> imageBuffer)
                 {
                     labelGetter(ref label);
                     imageGetter(ref image);
+                    if (image.Length <= 0)
+                        continue; //Empty Image
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: newline after this

File: src/Microsoft.ML.ImageAnalytics/ImageLoader.cs [hunk 2]
```diff
@@ -255,31 +254,30 @@ private Delegate MakeGetterVectorDataViewByteType(DataViewRow input, int iinfo,
...
                             string path = src.ToString();
                             if (!string.IsNullOrWhiteSpace(_parent.ImageFolder))
                                 path = Path.Combine(_parent.ImageFolder, path);
-                            if (!TryLoadDataIntoBuffer(path, ref dst, buffer))
-                                throw Host.Except($"Failed to load image {src.ToString()}.");
+                            if (!TryLoadDataIntoBuffer(path, ref dst))
+                            {
+                                var editor = VBufferEditor.Create(ref dst, 0); //Empty Image
+                                dst = editor.Commit();
+                            }
```
Comment (codemzs) [inline src/Microsoft.ML.ImageAnalytics/ImageLoader.cs]: Love it! This is the right way to handle invalid images and of course ImageClassificationAPI handles this case by dropping it

File: src/Microsoft.ML.ImageAnalytics/ImageLoader.cs [hunk 3]
```diff
@@ -302,28 +300,42 @@ private static bool TryLoadDataIntoBuffer(string path, ref VBuffer<byte> imgData
...
             {
                 int totalBytesRead = 0;
                 int bytesRead;
-                var readBufferSpan = readBuffer.AsSpan();
-                var srcSpan = readBufferSpan;
-                while ((bytesRead = stream.Read(readBuffer, 0, readBuffer.Length)) > 0)
+
+                int chunksize = 4096; // Most optimal size for buffer, friendly to CPU's L1 cache
+                var bufferPool = ArrayPool<byte>.Shared;
+                byte[] readBuffer = bufferPool.Rent(chunksize);
```
Comment (codemzs) [inline src/Microsoft.ML.ImageAnalytics/ImageLoader.cs]: >Rent [](start = 47, length = 4)

I prefer the term "Lease" and have seen that in software stack but first time seeing "Rent"

File: src/Microsoft.ML.ImageAnalytics/ImageLoader.cs [hunk 4]
```diff
@@ -302,28 +300,42 @@ private static bool TryLoadDataIntoBuffer(string path, ref VBuffer<byte> imgData
...
+                        var dstSpan = bufferSpan.Slice(totalBytesRead, bytesRead);
+                        Contract.Assert(srcSpan.Length == dstSpan.Length);
+                        srcSpan.CopyTo(dstSpan);
+                        totalBytesRead += bytesRead;
+                    }
                 }
+                finally
+                {
+                    bufferPool.Return(readBuffer);
+                    // don't use the reference to the buffer after returning it!
```
Comment (codemzs) [inline src/Microsoft.ML.ImageAnalytics/ImageLoader.cs]: >// don't use the reference to the buffer after returning it! [](start = 20, length = 60)

comment goes above the code it is referring

**Comment by codemzs** [comment]
@harshithapv Why is this a draft PR? It seems pretty good.

File: src/Microsoft.ML.ImageAnalytics/ImageLoader.cs (hunk 3 above)
Comment (eerhardt) [reply src/Microsoft.ML.ImageAnalytics/ImageLoader.cs]: `Rent` and `Return` are the terms chosen in the `System.Buffers` library. They are used both for ArrayPool, and MemoryPool.

[https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.rent](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.rent)
[https://docs.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1.rent](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1.rent)

---
In reply to: [331301719](https://github.com/dotnet/machinelearning/pull/4293#discussion_r331301719) [](ancestors = 331301719)

**Pull request merged**